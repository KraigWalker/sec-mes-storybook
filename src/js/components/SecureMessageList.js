import React from 'react';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import SecureMessageSummary from './common/SecureMessageSummary';
import { getMessageType } from '../utils/SecureMessageUtils';
import { sendMessageForAccessibiltiy } from '../actions/AppActions';
import { connect } from 'react-redux';
class SecureMessageList extends React.Component {
    constructor(props) {
        super(props);
        this.showMoreClicked = this.showMoreClicked.bind(this);
        this.state = {
            showMoreLimit: 20,
            showThatsAllMessage: false,
        }
    }
    componentWillReceiveProps(props) {
        if (this.props.messages.length <= 20) {
            this.setState({ showThatsAllMessage: true });
        } else {
            this.setState({ showMoreLimit: 20, showThatsAllMessage: '' });
        }

        if (this.props.activeTab === 'sent' && this.props.messagesFetched) {
            this.props.dispatch(sendMessageForAccessibiltiy('You don’t have any sent messages'));
        } else if (this.props.activeTab === 'drafts' && this.props.messagesFetched) {
            this.props.dispatch(sendMessageForAccessibiltiy('You haven’t saved any drafts'));
        } else if (this.props.activeTab === 'inbox' && this.props.messagesFetched) {
            //this.props.dispatch(sendMessageForAccessibiltiy('You don’t have any messages'));
        }
    }
    showMessages() {
        let msgs = this.props.messages.slice(0, this.state.showMoreLimit);
        const { messages } = this.props;
        const allMessages = [];
        const hasOnClick = true;
        const listFlag = true;
        _.map(msgs, (message, index) => {
            allMessages.push(
                <li key={index} className="c-messagelist__wrapper"><SecureMessageSummary message={message} listFlag={listFlag} /></li>)
        })
        return allMessages;
    }
    showMoreClicked() {

        this.props.dispatch(sendMessageForAccessibiltiy(`Next 20 messages loaded ${this.props.activeTab}`));
        let limit = this.props.messages.length;
        this.setState({
            showMoreLimit: limit,
            showThatsAllMessage: true,
        });
    }
    renderShowMoreButton() {
        let buttonText = 'Show more messages';
        if (this.props.activeTab === 'sent') {
            buttonText = 'Show more sent messages';
        } else if (this.props.activeTab === 'drafts') {
            buttonText = 'Show more drafts';
        }
        if (this.state.showMoreLimit < this.props.messages.length && (this.props.activeTab === 'sent' || this.props.activeTab === 'inbox' || this.props.activeTab === 'drafts')) {
            return (<button type="button" onClick={this.showMoreClicked} className="c-btn c-btn--default c-modal__button u-margin-bottom-c">{buttonText}</button>);
        }

    }
    renderThatsAllText() {
        let thatsallText = 'That’s all of them. Any new messages you send will appear at the top of this list. We display messages up to 13 months in the past.';
        if (this.props.activeTab === 'sent') {
            thatsallText = 'That’s all of them. Any new messages you send will appear at the top of this list. We display messages up to 13 months in the past.'
        } else if (this.props.activeTab === 'drafts') {
            thatsallText = 'That’s all of them. Any new drafts you save will appear at the top of this list. We display messages up to 13 months in the past.';
        }
        return thatsallText;
    }
    renderNoMessagesText() {
        if (this.props.activeTab === 'sent') {
            //  this.props.dispatch(sendMessageForAccessibiltiy('You don’t have any sent messages'));
            return (<li className="c-messagelist__wrapper">You don’t have any sent messages.</li>);
        } else if (this.props.activeTab === 'drafts') {
            // this.props.dispatch(sendMessageForAccessibiltiy('You haven’t saved any drafts'));
            return (<li className="c-messagelist__wrapper">You haven’t saved any drafts.</li>);
        } else {
            // this.props.dispatch(sendMessageForAccessibiltiy('You don’t have any messages'));
            return (<li className="c-messagelist__wrapper">You don’t have any messages.</li>);
        }
    }
    render() {

        return (
            <section>
                <ol className="c-messagelist">
                    {this.props.messages.length === 0 ? this.renderNoMessagesText() : this.showMessages()}
                </ol>
                {this.renderShowMoreButton()}
                {this.state.showThatsAllMessage && <p className="u-margin-bottom-c">{this.renderThatsAllText()}</p>}
            </section>
        );
    }
}

// export default SecureMessageList;
/**
 * Maps the state of the component to the state of the redux store
 * @param {object} state. State of the application
 */
const mapState = (state) => {
    return {
        messagesubjects: state.subjects,
        messageaccounts: state.accounts,
        messagesFetched: state.messages.fetched,
    }
};
export default connect(mapState)(SecureMessageList);

