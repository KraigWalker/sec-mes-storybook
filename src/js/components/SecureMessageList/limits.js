export const LIMITS = [20, 150, 500];

export const getNextLimit = ({ currentLimit, messageCount }) => {
  const nextLimitIdx = LIMITS.findIndex((l) => l === currentLimit) + 1;
  return LIMITS[nextLimitIdx] ? LIMITS[nextLimitIdx] : messageCount;
};

export const shouldShowThatsAllMessage = ({ currentLimit, messageCount }) =>
  messageCount !== 0 && messageCount <= currentLimit;
