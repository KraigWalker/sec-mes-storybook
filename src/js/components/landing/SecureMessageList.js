import React from 'react';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import SecureMessage from './SecureMessage';


class SecureMessageList extends React.Component {
    showMessages() {
        const { messages } = this.props;
        const allMessages = [];
        const hasOnClick = true;
        const listFlag = true;
        _.map(messages, (message, index) => {
            allMessages.push(<li className="c-messagelist__wrapper"><SecureMessage message= { message } key={index} hasOnClick={hasOnClick} listFlag={listFlag}/></li>)
        })
        return allMessages;
    }
    render() {
        return (
            <section>
                <ol className="c-messagelist">
                    {this.showMessages()}
                </ol>
            </section>
        );  
    }
}

export default SecureMessageList;

