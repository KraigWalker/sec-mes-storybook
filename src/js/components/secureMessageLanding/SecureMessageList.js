import React from 'react';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import SecureMessageSummary from '../common/SecureMessageSummary';


class SecureMessageList extends React.Component {
    showMessages() {
        const { messages } = this.props;
        const allMessages = [];
        const hasOnClick = true;
        const listFlag = true;
        _.map(messages, (message, index) => {
            allMessages.push(<li className="c-messagelist__wrapper"><SecureMessageSummary message= { message } key={index} hasOnClick={hasOnClick} listFlag={listFlag}/></li>)
        })
        return allMessages;
    }
    render() {
        return (
            <section>
                <ol className="c-messagelist">
                    {this.showMessages()}
                </ol>
            </section>
        );  
    }
}

export default SecureMessageList;

