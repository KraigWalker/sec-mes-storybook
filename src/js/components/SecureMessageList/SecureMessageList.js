import React from 'react';
import SecureMessageSummary from '../common/SecureMessageSummary';
import { sendMessageForAccessibiltiy as sendMessageForAccessibility } from '../../actions/AppActions';
import { connect } from 'react-redux';
import { SENT, DRAFT, ARCHIVED } from '../../constants/StringsConstants';
import { Button } from 'web-ui-components/lib/atoms/buttons';
import { Column } from 'web-ui-components/lib/global/layout';
import { TextBody } from 'web-ui-components/lib/atoms/text';
import { LoadingLocalTakeover } from 'web-ui-components/lib/organisms/takeovers';
import { TextStyled } from 'web-ui-components/lib/atoms/text';
import { Mail } from 'web-ui-components/lib/communication/messaging';
import { withBreakpoints } from '../common/hoc/WithBreakpoint';
import { compose } from 'redux';
import { MessageSelectors } from '../../reducers';
import { LIMITS, shouldShowThatsAllMessage, getNextLimit } from './limits';

export class SecureMessageList extends React.Component {
  constructor(props) {
    super(props);
    this.showMoreClicked = this.showMoreClicked.bind(this);
    this.sendAccessibilityMessage = this.sendAccessibilityMessage.bind(this);
    this.state = {
      currentMessageLimit: LIMITS[0],
      showThatsAllMessage: false,
    };
  }

  componentWillReceiveProps(props) {
    const { messages } = props;
    this.setState({
      showThatsAllMessage: shouldShowThatsAllMessage({
        currentLimit: this.state.currentMessageLimit,
        messageCount: messages.length,
      }),
    });
  }

  componentDidMount() {
    const { messages } = this.props;
    if (messages.length === 0) {
      this.sendAccessibilityMessage();
    }
  }

  showMessages() {
    const { messages, content } = this.props;
    const listFlag = true;
    return messages.slice(0, this.state.currentMessageLimit).map((message, index) => <SecureMessageSummary key={index} message={message} listFlag={listFlag} content={content} />);
  }

  showMoreClicked() {
    const { messages, dispatch, activeTab } = this.props;
    dispatch(sendMessageForAccessibility(`Next 20 messages loaded ${activeTab}`));
    this.setState(prevState => {
      const newLimit = getNextLimit({
        currentLimit: prevState.currentMessageLimit,
        messageCount: messages.length,
      });
      return {
        currentMessageLimit: newLimit,
        showThatsAllMessage: shouldShowThatsAllMessage({
          currentLimit: newLimit,
          messageCount: messages.length,
        }),
      };
    });
  }

  renderThatsAllText() {
    const { content, activeTab } = this.props;
    let thatsallText = content.thatsallTextInbox;
    if (activeTab === SENT) {
      thatsallText = content.thatsallTextSend;
    } else if (activeTab === ARCHIVED) {
      thatsallText = content.thatsallTextArchived;
    } else if (activeTab === DRAFT) {
      thatsallText = content.thatsallTextDraft;
    }
    return (
      <TextStyled size="uist" className="u-padding-top-2">
        {thatsallText}
      </TextStyled>
    );
  }

  renderNoMessagesText(isLoading) {
    const { content } = this.props;
    return <Mail.Empty className="u-padding-left-1">{isLoading ? 'loading' : content.noMessages}</Mail.Empty>;
  }

  sendAccessibilityMessage() {
    const { activeTab, dispatch, content } = this.props;
    switch (activeTab) {
      case SENT:
        dispatch(sendMessageForAccessibility(content.noSentMessages));
        break;
      case DRAFT:
        dispatch(sendMessageForAccessibility(content.noDraftMessages));
        break;
      case ARCHIVED:
        dispatch(sendMessageForAccessibility(content.noArchivedMessages));
        break;
      default:
        dispatch(sendMessageForAccessibility(content.noInboxMessages));
        break;
    }
  }

  render() {
    const { messagesFetching, messages, content, noPadding } = this.props;

    let paddingProps = null;
    if (noPadding) {
      paddingProps = {
        className: 'u-padding-left-1',
      };
    }

    return (
      <Column xs={24} {...paddingProps}>
        <LoadingLocalTakeover xs={24} show={messagesFetching} title="loading..">
          <TextBody>
            {messages.length === 0 ? this.renderNoMessagesText(messagesFetching) : this.showMessages()}
            {this.state.currentMessageLimit < messages.length && (
              <Button className="u-margin-top-2" display="primary" onClick={this.showMoreClicked}>
                {content.showMore}
              </Button>
            )}
            {this.state.showThatsAllMessage && this.renderThatsAllText()}
          </TextBody>
        </LoadingLocalTakeover>
      </Column>
    );
  }
}

// export default SecureMessageList;
/**
 * Maps the state of the component to the state of the redux store
 * @param {object} state. State of the application
 */
const mapState = state => ({
  messagesubjects: state.subjects,
  messageaccounts: state.accounts,
  messagesFetching: MessageSelectors.getFetching(state),
});

export default compose(
  connect(mapState),
  withBreakpoints
)(SecureMessageList);
