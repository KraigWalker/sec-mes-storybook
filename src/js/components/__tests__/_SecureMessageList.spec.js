import React from 'react';
import { SecureMessageList } from '../SecureMessageList';
import { shallow } from 'enzyme';
import TestData from '../../content/secureMessagesTestData.json'

describe("New Secure message snapshot", () => {
    const dispatch = jest.fn();
    const slice = jest.fn();
    let props = {
        content: {
            back: 'Back',
        },
        activeTab: 'SENT',
        messages: TestData.secure_messages,
        messagesFetched: {
            fetching: false,
            successModal: false
        },
        dispatch: dispatch
    };
    let component = shallow(<SecureMessageList {...props} />);
    it('should match to snapshot', () => {
        component.setState({ showThatsAllMessage: true });
        expect(component).toMatchSnapshot();
    });
    it('showMoreClicked function test', () => {
        component.instance().showMoreClicked();
        component.setState({ showThatsAllMessage: true });
        expect(component.instance().state.showThatsAllMessage).toBeTruthy();
    });
    it('renderNoMessagesText function test', () => {
        component.instance().renderNoMessagesText();
        component.setState({ showMoreLimit: 5 });
        expect(component.instance().state.showMoreLimit).toBe(props.messages.length);
    });
    describe("renderNoMessagesText with DRAFT", () => {
        const dispatch = jest.fn();
        const slice = jest.fn();
        let props = {
            content: {
                noDraftMessages: 'noDraftMessages',
                thatsallTextDraft: 'thatsallTextDraft'
            },
            activeTab: 'DRAFT',
            messagesFetched: {
                fetching: false,
                successModal: false
            },
            messages: {
                slice: slice,
                length: 10
            },
            dispatch: dispatch
        };
        it('renderNoMessagesText function test', () => {
            let component = shallow(<SecureMessageList {...props} />);
            component.instance().renderNoMessagesText();
            component.setState({ showMoreLimit: 8 });
            expect(component.instance().state.showMoreLimit).toBeLessThan(props.messages.length);
        });
    });
    describe("New Secure message snapshot", () => {
        const dispatch = jest.fn();
        const slice = jest.fn();
        let props = {
            content: {
                back: 'Back',
            },
            messages: {
                slice: slice
            },
            messagesFetched: {
                fetching: true,
                successModal: true
            },
            dispatch: dispatch
        };
        let component = shallow(<SecureMessageList {...props} />);
        it('should match to snapshot', () => {
            expect(component).toMatchSnapshot();
        });
    });
});
