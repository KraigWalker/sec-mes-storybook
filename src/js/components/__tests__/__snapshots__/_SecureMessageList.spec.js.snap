// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Secure message snapshot New Secure message snapshot should match to snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SecureMessageList
    content={
      Object {
        "back": "Back",
      }
    }
    dispatch={[MockFunction]}
    messages={
      Object {
        "slice": [MockFunction] {
          "calls": Array [
            Array [
              0,
              20,
            ],
          ],
        },
      }
    }
    messagesFetched={
      Object {
        "fetching": true,
        "successModal": true,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ol
          className="c-messagelist"
        />,
        undefined,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "c-messagelist",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ol",
      },
      undefined,
      false,
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ol
            className="c-messagelist"
          />,
          undefined,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "c-messagelist",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ol",
        },
        undefined,
        false,
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenFourAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`New Secure message snapshot should match to snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SecureMessageList
    activeTab="SENT"
    content={
      Object {
        "back": "Back",
      }
    }
    dispatch={[MockFunction]}
    messages={
      Object {
        "length": 10,
        "slice": [MockFunction] {
          "calls": Array [
            Array [
              0,
              20,
            ],
            Array [
              0,
              20,
            ],
          ],
        },
      }
    }
    messagesFetched={
      Object {
        "fetching": false,
        "successModal": false,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ol
          className="c-messagelist"
        />,
        undefined,
        <p
          className="u-margin-bottom-c"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "c-messagelist",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ol",
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "u-margin-bottom-c",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ol
            className="c-messagelist"
          />,
          undefined,
          <p
            className="u-margin-bottom-c"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "c-messagelist",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ol",
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "u-margin-bottom-c",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenFourAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
